src/dataflow/decode/mod.rs

```diff
@@ -118,6 +118,306 @@ where
     })
 }
 
+pub type PushSession<'a, R> =
+    Session<'a, Timestamp, R, PushCounter<Timestamp, R, Tee<Timestamp, R>>>;
+
+pub trait DecoderState {
+    /// Reset number of success and failures with decoding
+    fn reset_event_count(&mut self);
+    fn decode_key(&mut self, bytes: &[u8]) -> Result<Row, String>;
+    /// give a session a key-value pair
+    fn give_key_value<'a>(
+        &mut self,
+        key: Row,
+        bytes: &[u8],
+        aux_num: &Option<i64>,
+        session: &mut PushSession<'a, (Row, Option<Row>, Timestamp)>,
+        time: Timestamp,
+    );
+    /// give a session a plain value
+    fn give_value<'a>(
+        &mut self,
+        bytes: &[u8],
+        aux_num: &Option<i64>,
+        session: &mut PushSession<'a, (Row, Timestamp, Diff)>,
+        time: Timestamp,
+    );
+    /// Register number of success and failures with decoding
+    fn log_error_count(&self);
+}
+
+fn pack_with_line_no(datum: Datum, line_no: &Option<i64>) -> Row {
+    Row::pack(iter::once(datum).chain(line_no.map(Datum::from)))
+}
+
+fn bytes_to_datum(bytes: &[u8]) -> Datum {
+    Datum::from(bytes)
+}
+
+fn text_to_datum(bytes: &[u8]) -> Datum {
+    Datum::from(std::str::from_utf8(bytes).ok())
+}
+
+struct OffsetDecoderState<F: Fn(&[u8]) -> Datum> {
+    datum_func: F,
+}
+
+impl<F: Fn(&[u8]) -> Datum> DecoderState for OffsetDecoderState<F> {
+    fn reset_event_count(&mut self) {}
+
+    fn decode_key(&mut self, bytes: &[u8]) -> Result<Row, String> {
+        let mut result = RowPacker::new();
+        result.push((self.datum_func)(bytes));
+        Ok(result.finish())
+    }
+
+    /// give a session a key-value pair
+    fn give_key_value<'a>(
+        &mut self,
+        key: Row,
+        bytes: &[u8],
+        line_no: &Option<i64>,
+        session: &mut PushSession<'a, (Row, Option<Row>, Timestamp)>,
+        time: Timestamp,
+    ) {
+        session.give((
+            key,
+            Some(pack_with_line_no((self.datum_func)(bytes), line_no)),
+            time,
+        ));
+    }
+
+    /// give a session a plain value
+    fn give_value<'a>(
+        &mut self,
+        bytes: &[u8],
+        line_no: &Option<i64>,
+        session: &mut PushSession<'a, (Row, Timestamp, Diff)>,
+        time: Timestamp,
+    ) {
+        session.give((
+            pack_with_line_no((self.datum_func)(bytes), line_no),
+            time,
+            1,
+        ));
+    }
+
+    /// Register number of success and failures with decoding
+    fn log_error_count(&self) {}
+}
+
+fn decode_kafka_upsert_inner<G, K: 'static, V: 'static>(
+    stream: &Stream<G, ((Vec<u8>, Vec<u8>), Option<i64>)>,
+    mut key_decoder_state: K,
+    mut value_decoder_state: V,
+    op_name: &str,
+) -> Stream<G, (Row, Option<Row>, Timestamp)>
+where
+    G: Scope<Timestamp = Timestamp>,
+    K: DecoderState,
```

Comment:
It is a bit confusing to have trait bounds both in the declaration and in the where clause for the same trait. I would prefer

```
fn decode_kafka_upsert_inner<G, K, V>
(...)
where
      G: Scope<Timestamp = Timestamp>,
      K: DecoderState + 'static,
      V: DecoderState + 'static
```